plugin/db/mysql/restore.go

```diff
@@ -261,127 +273,70 @@ func sortBinlogFiles(binlogFiles []BinlogFile) []BinlogFile {
 			continue
 		}
 		validBackupList = append(validBackupList, b)
-		eventTs, err := r.parseLocalBinlogEventTimestamp(ctx, b.Payload.BinlogInfo)
-		if err != nil {
-			return nil, fmt.Errorf("failed to parse binlog event timestamp, error: %w", err)
-		}
-		eventTsList = append(eventTsList, eventTs)
 	}
-	log.Debug("Binlog event ts list of backups", zap.Int64s("eventTsList", eventTsList))
-
-	backup, err := getLatestBackupBeforeOrEqualTsImpl(validBackupList, eventTsList, targetTs)
-	if err != nil {
-		return nil, err
-	}
-	return backup, nil
 
+	return getLatestBackupBeforeOrEqualBinlogCoord(validBackupList, *targetBinlogCoordinate)
 }
 
-// The backupList must 1 to 1 maps to the eventTsList, and the sorting order is not required.
-func getLatestBackupBeforeOrEqualTsImpl(backupList []*api.Backup, eventTsList []int64, targetTs int64) (*api.Backup, error) {
-	var maxEventTsLETargetTs int64
-	var minEventTs int64 = math.MaxInt64
-	var backup *api.Backup
-	for i, b := range backupList {
-		// Parse the binlog files and convert binlog positions into MySQL server timestamps.
-		if b.Payload.BinlogInfo.IsEmpty() {
-			continue
-		}
-		eventTs := eventTsList[i]
-		if eventTs <= targetTs && eventTs > maxEventTsLETargetTs {
-			maxEventTsLETargetTs = eventTs
-			backup = b
-		}
-		// This is only for composing the error message when no valid backup found.
-		if eventTs < minEventTs {
-			minEventTs = eventTs
+func getLatestBackupBeforeOrEqualBinlogCoord(backupList []*api.Backup, targetBinlogCoordinate binlogCoordinate) (*api.Backup, error) {
+	var backupCoordinateList []binlogCoordinate
+	for _, b := range backupList {
+		c, err := newBinlogCoordinate(b.Payload.BinlogInfo.FileName, b.Payload.BinlogInfo.Position)
+		if err != nil {
+			return nil, err
 		}
+		backupCoordinateList = append(backupCoordinateList, c)
 	}
 
-	if maxEventTsLETargetTs == 0 {
-		targetDateTime := time.Unix(targetTs, 0).Format(time.RFC822)
-		minEventDateTime := time.Unix(minEventTs, 0).Format(time.RFC822)
-		log.Debug("the target restore time is earlier than the oldest backup time",
-			zap.String("targetDatetime", targetDateTime),
-			zap.Int64("targetTimestamp", targetTs),
-			zap.String("minEventDateTime", minEventDateTime),
-			zap.Int64("minEventTimestamp", minEventTs))
+	var maxSeq int64
+	var maxPos int64
+	var backup *api.Backup
+	for i, c := range backupCoordinateList {
+		if c.Seq < targetBinlogCoordinate.Seq || (c.Seq == targetBinlogCoordinate.Seq && c.Pos <= targetBinlogCoordinate.Pos) {
+			if c.Seq > maxSeq || (c.Seq == maxSeq && c.Pos > maxPos) {
+				maxSeq = c.Seq
+				maxPos = c.Pos
+				backup = backupList[i]
+			}
+		}
+	}
 
-		return nil, fmt.Errorf("the target restore time %s is earlier than the oldest backup time %s", targetDateTime, minEventDateTime)
+	if backup == nil {
+		var minSeq int64 = math.MaxInt64
+		var minPos int64 = math.MaxInt64
+		var oldestBackupBinlogCoordinate binlogCoordinate
+		for _, c := range backupCoordinateList {
```

Comment:
If we sort the backupCoordinateList, the math will be easier.

Reply:
SG.